### PR TITLE
Added in  reflect Select.overloaded parameter for expected result type [WIP]

### PR DESCRIPTION
**Pull request by rssh**
Motivation is  described in #10053

### CODE REVIEW

**Comment by bishabosha** [comment]
would it make more sense to add a method that can call with full curried parameter lists?

**Comment by rssh** [comment]
From the point of usability, this can be better,  but overloading is defined in language specs only for one parameter list. (https://scala-lang.org/files/archive/spec/2.13/06-expressions.html#overloading-resolution ),  I guess because it's a beast from the java world.  

(after reading specs:  looks like things like:
```
def f(x:A)(y:B)
def f(x:A)(y:C) 
```
  should not compile at all.  But they compiled.

The compiler now counts the weight of the transformations for choosing the best alternative only for the arguments in the first list and uses the second argument list only for return type constraint.

 If we extending the notion of overloading to multiple argument lists, then we probably should change the logic of the compiler, to use other lists during the generation of alternatives.

So, the question is in specs:  I see 3 variants of resolution
  -   extend overloading to the list of arguments, change specs, and semantics of resolving overloads.  In this case List of list of arguments in API  is logical. 
  -  assume that overloading is defined only for the first argument list and other uses as constraints for the return type,  add a description of participating return type in alternatives to specs and return API as in this request.
  - if we have the next major version, maybe just disallow using  return type [write this in specs] and   API should not contain anything behind the first argument list  (i.e. as now, without this pull request).

Interesting question, I don't know the answer.

**Comment by bishabosha** [comment]
My suggestion was to keep calling the `tpd.applyOverloaded` method but for the resultType use a `MethodType` recursively built from the following parameter lists

**Comment by rssh** [comment]
yes, .. I thinking  this can be viewed as priorities. [i.e. choose the first list of arguments by weight, than second  ]. ... but not sure exactly,

I will prepare a variant with the list.

**Comment by rssh** [comment]
Problem with list of list approach: if we assume that the last type should always be a wildcard, then we can't be able to compile our example:  compiler assumes that   ```[PartialFunction[A,B]=>_]```   and ```[A=>B]=>_``` both satisfy the same constraint.  Therefore next code will fail:


```
package x

object X:

  def andThen[A,B](a:A)(f: A => B): B =
     println("totalFunction")
     f(a)

  def andThen[A,B](a:A)(f: PartialFunction[A,B]): Option[B] =
     println("partialFunction")
     f.lift.apply(a)

object Main:

 def main(args:Array[String]):Unit =
    // compiled
    val x1 = X.andThen(1){case x if (x%2 == 0) => x}
    val x2 = Macro.mThen{case x:Int if (x%2 == 0) => x}
```

```
package x

import scala.quoted._

object Macro:

    inline def mThen[A,B](inline x:A=>B):B = ${
       mThenImpl[A,B,A=>B,B]('x)
    }

    inline def mThen[A,B](inline x:PartialFunction[A,B]): Option[B] = ${
       mThenImpl[A,B,PartialFunction[A,B],Option[B]]('x)
    }

    def mThenImpl[A:Type, B:Type, S<:(A=>B) :Type, R:Type](x:Expr[S])(using qctx: QuoteContext):Expr[R]=
       import qctx.reflect._
       val fun = '{X}.unseal
       //val wildcard = TypeBounds(quoted.Type[Nothing].unseal.tpe, quoted.Type[Any].unseal.tpe)
       val wildcard = TypeBounds(TypeIdent(defn.NothingClass).tpe, TypeIdent(defn.AnyClass).tpe)
       val firstPart = Select.overloaded(fun,"andThen",
                                 List(TypeIdent(defn.IntClass).tpe, TypeIdent(defn.IntClass).tpe),
                                 List(Literal(Constant.Int(1))),
                                 quoted.Type[(S) => _].unseal.tpe
                       )
       Apply(firstPart,List(x.unseal)).seal.cast[R]

~   
```

File: library/src/scala/tasty/Reflection.scala [hunk 1]
```diff
@@ -2297,6 +2297,7 @@ trait Reflection { reflection =>
   trait TypeBoundsModule { this: TypeBounds.type =>
     def apply(low: Type, hi: Type): TypeBounds
     def unapply(x: TypeBounds): Option[(Type, Type)]
+    def minmax: TypeBounds
```
Comment (nicolasstucki) [inline library/src/scala/tasty/Reflection.scala:2300]:
```suggestion
    def empty: TypeBounds
    def upper(hi: Type): TypeBounds
    def lower(lo: Type): TypeBounds
```

File: tests/neg-staging/i5941/macro_1.scala [hunk 2]
```diff
@@ -17,7 +17,7 @@ object Lens {
     import util._
     // obj.copy(field = value)
     def setterBody(obj: Expr[S], value: Expr[T], field: String): Expr[S] =
-      Select.overloaded(obj.unseal, "copy", Nil, NamedArg(field, value.unseal) :: Nil).seal.cast[S]
+      Select.overloaded(obj.unseal, "copy", Nil, NamedArg(field, value.unseal) :: Nil, TypeBounds.minmax).seal.cast[S]
```
Comment (nicolasstucki) [inline tests/neg-staging/i5941/macro_1.scala:20]:
```suggestion
      Select.overloaded(obj.unseal, "copy", Nil, NamedArg(field, value.unseal) :: Nil, TypeBounds.empty).seal.cast[S]
```

File: library/src/scala/tasty/Reflection.scala [hunk 3]
```diff
@@ -514,7 +514,7 @@ trait Reflection { reflection =>
 
     // TODO rename, this returns an Apply and not a Select
     /** Call an overloaded method with the given type and term parameters */
-    def overloaded(qualifier: Term, name: String, targs: List[Type], args: List[Term]): Apply
+    def overloaded(qualifier: Term, name: String, targs: List[Type], args: List[Term], returnType: Type): Apply
```
Comment (nicolasstucki) [inline library/src/scala/tasty/Reflection.scala:517]:
```suggestion
    def overloaded(qualifier: Term, name: String, targs: List[Type], args: List[Term]): Apply
      
    /** Call an overloaded method with the given type and term parameters */
    def overloaded(qualifier: Term, name: String, targs: List[Type], args: List[Term], returnType: Type): Apply
```

File: tests/pos-macros/i6535/Macro_1.scala [hunk 4]
```diff
@@ -12,7 +12,7 @@ object scalatest {
       case t @ Apply(Select(lhs, op), rhs :: Nil) =>
         let(lhs) { left =>
           let(rhs) { right =>
-            val app = Select.overloaded(left, op, Nil, right :: Nil)
+            val app = Select.overloaded(left, op, Nil, right :: Nil, TypeBounds.minmax)
```
Comment (nicolasstucki) [inline tests/pos-macros/i6535/Macro_1.scala:15]:
```suggestion
            val app = Select.overloaded(left, op, Nil, right :: Nil)
```

File: tests/run-macros/i5941/macro_1.scala [hunk 5]
```diff
@@ -19,7 +19,7 @@ object Lens {
     def setterBody(obj: Term, value: Term, parts: List[String]): Term = {
       // o.copy(field = value)
       def helper(obj: Term, value: Term, field: String): Term =
-        Select.overloaded(obj, "copy", Nil, NamedArg(field, value) :: Nil)
+        Select.overloaded(obj, "copy", Nil, NamedArg(field, value) :: Nil, TypeBounds.minmax)
```
Comment (nicolasstucki) [inline tests/run-macros/i5941/macro_1.scala:22]:
```suggestion
        Select.overloaded(obj, "copy", Nil, NamedArg(field, value) :: Nil)
```

File: tests/run-macros/i5941/macro_1.scala [hunk 6]
```diff
@@ -118,7 +118,7 @@ object Iso {
       // (p: S) => p._1
       val to = (p: S) =>  ${ Select.unique(('p).unseal, "_1").seal.cast[A] }
       // (p: A) => S(p)
-      val from = (p: A) =>  ${ Select.overloaded(Ident(companion), "apply", Nil, ('p).unseal :: Nil).seal.cast[S] }
+      val from = (p: A) =>  ${ Select.overloaded(Ident(companion), "apply", Nil, ('p).unseal :: Nil, TypeBounds.minmax).seal.cast[S] }
```
Comment (nicolasstucki) [inline tests/run-macros/i5941/macro_1.scala:121]:
```suggestion
      val from = (p: A) =>  ${ Select.overloaded(Ident(companion), "apply", Nil, ('p).unseal :: Nil).seal.cast[S] }
```

File: tests/run-macros/i5941/macro_1.scala [hunk 7]
```diff
@@ -147,7 +147,7 @@ object Iso {
         case TypeRef(prefix, name) => TermRef(prefix, name)
       }
 
-      val obj = Select.overloaded(Ident(companion), "apply", Nil, Nil).seal.cast[S]
+      val obj = Select.overloaded(Ident(companion), "apply", Nil, Nil, TypeBounds.minmax).seal.cast[S]
```
Comment (nicolasstucki) [inline tests/run-macros/i5941/macro_1.scala:150]:
```suggestion
      val obj = Select.overloaded(Ident(companion), "apply", Nil, Nil).seal.cast[S]
```

File: tests/run-macros/i6171/Macro_1.scala [hunk 8]
```diff
@@ -16,7 +16,7 @@ object scalatest {
       case t @ Apply(Select(lhs, op), rhs :: Nil) =>
         let(lhs) { left =>
           let(rhs) { right =>
-            val app = Select.overloaded(left, op, Nil, right :: Nil)
+            val app = Select.overloaded(left, op, Nil, right :: Nil, TypeBounds.minmax)
```
Comment (nicolasstucki) [inline tests/run-macros/i6171/Macro_1.scala:19]:
```suggestion
            val app = Select.overloaded(left, op, Nil, right :: Nil)
```

File: tests/run-macros/i6171/Macro_1.scala [hunk 9]
```diff
@@ -30,7 +30,7 @@ object scalatest {
         if isImplicitMethodType(f.tpe) =>
         let(lhs) { left =>
           let(rhs) { right =>
-            val app = Select.overloaded(Apply(qual, left :: Nil), op, Nil, right :: Nil)
+            val app = Select.overloaded(Apply(qual, left :: Nil), op, Nil, right :: Nil, TypeBounds.minmax)
```
Comment (nicolasstucki) [inline tests/run-macros/i6171/Macro_1.scala:33]:
```suggestion
            val app = Select.overloaded(Apply(qual, left :: Nil), op, Nil, right :: Nil)
```

File: tests/run-macros/reflect-pos-fun/assert_1.scala [hunk 10]
```diff
@@ -12,7 +12,7 @@ object scalatest {
       case t @ Apply(TypeApply(Select(lhs, op), targs), rhs) =>
         let(lhs) { left =>
           lets(rhs) { rs =>
-            val app = Select.overloaded(left, op, targs.map(_.tpe), rs)
+            val app = Select.overloaded(left, op, targs.map(_.tpe), rs, TypeBounds.minmax)
```
Comment (nicolasstucki) [inline tests/run-macros/reflect-pos-fun/assert_1.scala:15]:
```suggestion
            val app = Select.overloaded(left, op, targs.map(_.tpe), rs)
```

File: tests/run-macros/reflect-select-constructor/assert_1.scala [hunk 11]
```diff
@@ -16,7 +16,7 @@ object scalatest {
       case t @ Apply(Select(lhs, op), rhs :: Nil) =>
         let(lhs) { left =>
           let(rhs) { right =>
-            val app = Select.overloaded(left, op, Nil, right :: Nil)
+            val app = Select.overloaded(left, op, Nil, right :: Nil, TypeBounds.minmax)
```
Comment (nicolasstucki) [inline tests/run-macros/reflect-select-constructor/assert_1.scala:19]:
```suggestion
            val app = Select.overloaded(left, op, Nil, right :: Nil)
```

File: tests/run-macros/reflect-select-constructor/assert_1.scala [hunk 12]
```diff
@@ -30,7 +30,7 @@ object scalatest {
       if isImplicitMethodType(f.tpe) =>
         let(lhs) { left =>
           let(rhs) { right =>
-            val app = Select.overloaded(Apply(qual, left :: Nil), op, Nil, right :: Nil)
+            val app = Select.overloaded(Apply(qual, left :: Nil), op, Nil, right :: Nil, TypeBounds.minmax)
```
Comment (nicolasstucki) [inline tests/run-macros/reflect-select-constructor/assert_1.scala:33]:
```suggestion
            val app = Select.overloaded(Apply(qual, left :: Nil), op, Nil, right :: Nil)
```

File: tests/run-macros/reflect-select-value-class/assert_1.scala [hunk 13]
```diff
@@ -16,7 +16,7 @@ object scalatest {
       case t @ Apply(Select(lhs, op), rhs :: Nil) =>
         let(lhs) { left =>
           let(rhs) { right =>
-            val app = Select.overloaded(left, op, Nil, right :: Nil)
+            val app = Select.overloaded(left, op, Nil, right :: Nil, TypeBounds.minmax)
```
Comment (nicolasstucki) [inline tests/run-macros/reflect-select-value-class/assert_1.scala:19]:
```suggestion
            val app = Select.overloaded(left, op, Nil, right :: Nil)
```

File: tests/run-macros/reflect-select-value-class/assert_1.scala [hunk 14]
```diff
@@ -30,7 +30,7 @@ object scalatest {
       if isImplicitMethodType(f.tpe) =>
         let(lhs) { left =>
           let(rhs) { right =>
-            val app = Select.overloaded(Apply(qual, left :: Nil), op, Nil, right :: Nil)
+            val app = Select.overloaded(Apply(qual, left :: Nil), op, Nil, right :: Nil, TypeBounds.minmax)
```
Comment (nicolasstucki) [inline tests/run-macros/reflect-select-value-class/assert_1.scala:33]:
```suggestion
            val app = Select.overloaded(Apply(qual, left :: Nil), op, Nil, right :: Nil)
```

File: compiler/src/dotty/tools/dotc/quoted/QuoteContextImpl.scala [hunk 15]
```diff
@@ -2049,6 +2049,7 @@ class QuoteContextImpl private (ctx: Context) extends QuoteContext:
     object TypeBounds extends TypeBoundsModule:
       def apply(low: Type, hi: Type): TypeBounds = Types.TypeBounds(low, hi)
       def unapply(x: TypeBounds): Option[(Type, Type)] = Some((x.low, x.hi))
+      def minmax = TypeBounds(defn.NothingClass.tree.tpe, defn.AnyClass.tree.tpe)
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/quoted/QuoteContextImpl.scala:2052]:
```suggestion
      def empty: TypeBounds = Types .TypeBounds.empty
      def upper(hi: Type): TypeBounds = Types .TypeBounds.upper(hi)
      def lower(lo: Type): TypeBounds = Types .TypeBounds.lower(hi)
```

File: compiler/src/dotty/tools/dotc/quoted/QuoteContextImpl.scala [hunk 16]
```diff
@@ -412,8 +412,8 @@ class QuoteContextImpl private (ctx: Context) extends QuoteContext:
         val denot = qualifier.tpe.member(name.toTermName)
         assert(!denot.isOverloaded, s"The symbol `$name` is overloaded. The method Select.unique can only be used for non-overloaded symbols.")
         withDefaultPos(tpd.Select(qualifier, name.toTermName))
-      def overloaded(qualifier: Term, name: String, targs: List[Type], args: List[Term]): Apply =
-        withDefaultPos(tpd.applyOverloaded(qualifier, name.toTermName, args, targs, Types.WildcardType).asInstanceOf[Apply])
+      def overloaded(qualifier: Term, name: String, targs: List[Type], args: List[Term], returnType: Type): Apply =
```
Comment (nicolasstucki) [inline compiler/src/dotty/tools/dotc/quoted/QuoteContextImpl.scala:415]:
```suggestion
      def overloaded(qualifier: Term, name: String, targs: List[Type], args: List[Term]): Apply =
        withDefaultPos(tpd.applyOverloaded(qualifier, name.toTermName, args, targs, Types.WildcardType).asInstanceOf[Apply])
        
      def overloaded(qualifier: Term, name: String, targs: List[Type], args: List[Term], returnType: Type): Apply =
```

File: tests/run-macros/tasty-overload-secondargs/Test_2.scala [hunk 17]
```diff
@@ -0,0 +1,12 @@
+
```
Comment (nicolasstucki) [inline tests/run-macros/tasty-overload-secondargs/Test_2.scala:1]: Nitpick: Could you remove the extra whitespaces on the top and bottom of this test?

**Review by nicolasstucki — COMMENTED** [review]
The fix seems good. Just needs some changes in the implementation within QuoteContextImpl. I also suggest to add the high and lower bound methods.

File: tests/run-macros/tasty-overload-secondargs/Test_2.scala (hunk 17 above)
Comment (rssh) [reply tests/run-macros/tasty-overload-secondargs/Test_2.scala]: dome

**Review by nicolasstucki — APPROVED** [review]

**Comment by nicolasstucki** [comment]
Thank you @rssh

**Pull request merged**